File: exbar.cpp
#include<vector>
#include<algorithm>
#include<chrono>
#include<climits>
#include <unordered_map>
#include "apta.h"
using namespace std;

INITIALIZE_EASYLOGGINGPP

void walkit(int redNode, int blueNode, int &nofChanges) {
    if (apta.nodes[blueNode]->label != LABEL_NONE) {
        if (apta.nodes[redNode]->label != LABEL_NONE) {
            if (apta.nodes[redNode]->label != apta.nodes[blueNode]->label) {
                throw 0; // returns to caller of `try_merge()`
            }
        } else {
            nofChanges += set_label(redNode, apta.nodes[blueNode]->label); // copy label
        }
    }
    for (int i = 0; i < sofAlphabet; ++i) {
        if (apta.nodes[blueNode]->children[i] != NO_CHILD) {
            if (apta.nodes[redNode]->children[i] != NO_CHILD) {
                walkit(apta.nodes[redNode]->children[i], apta.nodes[blueNode]->children[i], nofChanges); // recurse
            } else {
                nofChanges += set_child(redNode, i, apta.nodes[blueNode]->children[i]); // splice branch
            }
        }
    }
}

bool try_merge(int redNode, int blueNode, int &nofChanges) {
    nofChanges += set_father_point(redNode, blueNode); // creates loops
    try {
        walkit(redNode, blueNode, nofChanges);
    } catch (int e) {
        return false;
    }
    return true;
}

unordered_map<string, int> pickBlueNodeCache; 

int pickBlueNode() {
    vector<int> redNodes;
    vector<int> blueNodes;
    string currentDfaHash = "";
    for (int i = 0; i < apta.nodes.size(); ++i) {
        if (apta.nodes[i]->color == COLOR_RED) {
            redNodes.push_back(i);
            currentDfaHash += string("#") + to_string(i);
            for (int j = 0; j < apta.nodes[i]->children.size(); ++j) {
                int child = apta.nodes[i]->children[j];
                currentDfaHash += string("|") + to_string(child);
                if (
                    child != NO_CHILD
                    && apta.nodes[child]->color == COLOR_NONE
                    && apta.nodes[child]->isMerged == false
                    && find(blueNodes.begin(), blueNodes.end(), child) == blueNodes.end()
                ) {
                    blueNodes.push_back(child);
                }
            }
        }
    }
    if (blueNodes.size() == 0) {
        return -1;
    }
    /*unordered_map<std::string, int>::const_iterator got = pickBlueNodeCache.find(currentDfaHash);
    if (got != pickBlueNodeCache.end()) {
        LOG(DEBUG) << "The blue node has been read from the cache";
        return got->second;
    }*/
    int bestNofPossibleMerges = INT_MAX;
    int bestBlueNode = -1;
    for (int i = 0; i < blueNodes.size(); ++i) {
        int nofPossibleMerges = 0;
        for (int j = 0; j < redNodes.size(); ++j) {
            int nofChanges = 0;
            if (try_merge(redNodes[j], blueNodes[i], nofChanges)) {
                nofPossibleMerges++;
            }
            undo_changes(nofChanges);
        }
        LOG(DEBUG) << "The number of possible merge for node #" << blueNodes[i] << ": " << nofPossibleMerges;
        if (nofPossibleMerges < bestNofPossibleMerges) {
            bestBlueNode = blueNodes[i];
            bestNofPossibleMerges = nofPossibleMerges;
        }
    }
    //pickBlueNodeCache[currentDfaHash] = bestBlueNode; // saves the best blue node for the current DFA
    return bestBlueNode;
}

int max_red = 1;

void exh_search() {
    if (getNumberOfRedNodes() > max_red) {
        return;
    }
    LOG(DEBUG) << "The limit of red nodes = " << max_red;
    LOG(DEBUG) << "Looking for the best possible blue node...";
    int blueNode = pickBlueNode();
    LOG(DEBUG) << "The blue node has been picked: " << blueNode;
    if (blueNode != -1) {
        for (int i = 0; i < apta.nodes.size(); ++i) {
            if (apta.nodes[i]->color == COLOR_RED) {
                int nofChanges = 0;
                if (try_merge(i, blueNode, nofChanges)) {
                    LOG(DEBUG) << "Trying merging the blue node = " << blueNode << " and the red node = " << i << ": success";
                    nofChanges += set_as_merged(blueNode);
                    exh_search();
                } else {
                    LOG(DEBUG) << "Trying merging the blue node = " << blueNode << " and the red node = " << i << ": failure";
                }
                undo_changes(nofChanges);
            }
        }
        LOG(DEBUG) << "The blue node = " << blueNode << " has been transformed into a red node";
        set_color(blueNode, COLOR_RED);
        exh_search();
        undo_changes(1); // reverts the last red node transformation
    } else {
        throw 1;
    }
}

int main(int argc, char* argv[])
{
    START_EASYLOGGINGPP(argc, argv);
    el::Loggers::reconfigureAllLoggers(el::ConfigurationType::Format, "[%datetime] %msg");
    // the algorithm starts here
    string filepath = getCmdOption(argv, argv + argc, "--file");
    max_red = atoi(getCmdOption(argv, argv + argc, "--max"));
    if (max_red < 1) {
        max_red = 1;
    }
    build_APTA_from_file(filepath);
    LOG(INFO) << "[ exbar ]";
    //---
    clock_t time = clock();
    apta.nodes[0]->color = COLOR_RED;
    int iteration = 0;
    while (true) {
        try {
            LOG(INFO) << "The begin of the iteration = " << ++iteration << ", the limit of red nodes = " << max_red;
            exh_search();
            LOG(INFO) << "The end of the iteration = " << iteration;
            //getchar(); // test
            max_red++;
        } catch (int e) {
            LOG(INFO) << "The solution has been found";
            break;
        }
    }
    time = clock() - time;
    int ms = time / CLOCKS_PER_SEC * 1000;
    //---
    LOG(INFO) << "[ RESULT ]";
    LOG(INFO) << "The measured time in ms = " << ms;
    print_dfa();
    //---
    LOG(INFO) << "Saving the generated DFA to files";
    std::ofstream out1("dfa.txt");
    out1 << get_dfa(false);
    out1.close();
    std::ofstream out2("dfa.yaml");
    out2 << get_yaml_dfa();
    out2.close();
    //---
    if (cmdOptionExists(argv, argv+argc, "--verify")) {
        verify_APTA_from_file(filepath);
    }
    return 0;
}

File: exbar.php
<?php
$filepath = isset($argv[1]) ? $argv[1] : __DIR__ . '/samples/sample1.txt';
if (file_exists($filepath) === false || is_readable($filepath) === false) {
    exit("The filepath cannot be opened or read\n");
}

define('COLOR_NONE', 0);
define('COLOR_RED', 1);
define('COLOR_VIOLET', 2);
define('LABEL_NONE', 0);
define('LABEL_ACCEPTED', 1);
define('LABEL_REJECTED', -1);
define('NO_CHILD', -1);

/* This part reads and generates the APTA tree from the file. */

$apta = [];
$flags = FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES;
$lines = array_map(function ($line) {
    return array_map(function ($number) { return (int) $number; }, explode(' ', $line));
}, file($filepath, $flags));
$nofExamples = $lines[0][0];
$sofAlphabet = $lines[0][1];
// creates the apta tree root
$apta[0] = [
    'color' => COLOR_RED,
    'label' => LABEL_NONE,
    'children' => array_fill(0, $sofAlphabet, NO_CHILD)
];
$nodeIdx = 0;
foreach (array_slice($lines, 1) as $example) {
    $current = 0;
    for ($i = 0; $i < $example[1]; ++$i) {
        if ($apta[$current]['children'][$example[$i + 2]] === NO_CHILD) {
            $apta[++$nodeIdx] = [
                'color' => COLOR_NONE,
                'label' => LABEL_NONE,
                'children' => array_fill(0, $sofAlphabet, NO_CHILD)
            ];
            $apta[$current]['children'][$example[$i + 2]] = $nodeIdx;
        }
        $current = $apta[$current]['children'][$example[$i + 2]];
    }
    $apta[$current]['label'] = $example[0] ? LABEL_ACCEPTED : LABEL_REJECTED;
}

/* This part contains helper functions */

$nofPossibleMerges = [];
$changesHistory = [];

function getNofRedNodes() {
    global $apta;
    return array_reduce($apta, function ($carry, $node) {
        return $carry + ($node['color'] === COLOR_RED ? 1 : 0);
    }, 0);
}

function getNofPossibleMerges($label) {
    global $apta;
    return array_reduce($apta, function ($carry, $node) use ($label) {
        return $carry + ($node['color'] === COLOR_RED && ($node['label'] === LABEL_NONE || $node['label'] === $label) ? 1 : 0);
    }, 0);
}

function updateNofPossibleMerges() {
    global $nofPossibleMerges;
    foreach ([LABEL_NONE, LABEL_ACCEPTED, LABEL_REJECTED] as $label) {
        $nofPossibleMerges[$label] = getNofPossibleMerges($label);
    }
}

function setLabel($nodeIdx, $label) {
    global $apta, $changesHistory;
    $changesHistory[] = ['label', $nodeIdx, $apta[$nodeIdx]['label']];
    $apta[$nodeIdx]['label'] = $label;
}

function setColor($nodeIdx, $color) {
    global $apta, $changesHistory;
    $changesHistory[] = ['color', $nodeIdx, $apta[$nodeIdx]['color']];
    $apta[$nodeIdx]['color'] = $color;
}

function setChild($nodeIdx, $childIdx, $child) {
    global $apta, $changesHistory;
    $changesHistory[] = ['child', $nodeIdx, $childIdx, $apta[$nodeIdx]['children'][$childIdx]];
    $apta[$nodeIdx]['children'][$childIdx] = $child;
}

function setFatherPoint($blueNode, $redNode) {
    global $apta, $sofAlphabet;
    for ($i = 0; $i < count($apta); ++$i) {
        for ($j = 0; $j < $sofAlphabet; ++$j) {
            if ($apta[$i]['children'][$j] === $blueNode) {
                setChild($i, $j, $redNode);
            }
        }
    }
}

function revertChanges($limit = 1) {
    global $apta, $changesHistory;
    while (--$limit >= 0) {
        $change = array_pop($changesHistory);
        if ($change[0] === 'label') {
            $apta[$change[1]]['label'] = $change[2];
        } else if ($change[0] === 'child') {
            $apta[$change[1]]['children'][$change[2]] = $change[3];
        } else if ($change[0] === 'color') {
            $apta[$change[1]]['color'] = $change[2];
        }
    }
}

function indexNodes($nodeIdx, &$index, &$visited, &$map, &$acceptedStates, &$rejectedStates) {
    global $apta, $sofAlphabet;
    $map[$index] = $nodeIdx;
    if ($apta[$nodeIdx]['label'] === LABEL_ACCEPTED) {
        $acceptedStates[] = $index;
    }
    if ($apta[$nodeIdx]['label'] === LABEL_REJECTED) {
        $rejectedStates[] = $index;
    }
    $apta[$nodeIdx]['index'] = $index++;
    $visited[$nodeIdx] = $nodeIdx;
    for ($i = 0; $i < $sofAlphabet; ++$i) {
        if (
            $apta[$nodeIdx]['children'][$i] !== NO_CHILD
            && array_key_exists($apta[$nodeIdx]['children'][$i], $visited) === false
        ) {
            indexNodes($apta[$nodeIdx]['children'][$i], $index, $visited, $map, $acceptedStates, $rejectedStates);
        }
    }
}

function getDfa($result) {
    global $apta, $sofAlphabet;
    $dfa = $result['nofNodes'] . "\n";
    $dfa .= $sofAlphabet . "\n";
    $dfa .= trim(count($result['acceptedStates']) . ' ' . implode(' ', $result['acceptedStates'])) . "\n";
    $dfa .= trim(count($result['rejectedStates']) . ' ' . implode(' ', $result['rejectedStates'])) . "\n";
    foreach ($result['map'] as $index => $nodeIdx) {
        for ($i = 0; $i < $sofAlphabet; ++$i) {
            $dfa .= ($i > 0 ? ' ' : '') . ($apta[$nodeIdx]['children'][$i] === NO_CHILD ? NO_CHILD : $apta[$apta[$nodeIdx]['children'][$i]]['index']);
        }
        $dfa .= "\n";
    }
    return $dfa;
}

/* This part contains the EXBAR algorithm */

$maxNofRedNodes = 1;

function pickBlueNode($cutoff = 0) {
    global $apta, $nofPossibleMerges;
    $blueNodes = array_unique(array_reduce($apta, function ($carry, $node) use ($apta) {
        return array_merge($carry, $node['color'] !== COLOR_RED ? [] : array_filter($node['children'], function ($child) use ($apta) {
            return $child !== NO_CHILD && $apta[$child]['color'] === COLOR_NONE;
        }));
    }, []));
    echo '[' . implode(', ' , $blueNodes) . ']' . PHP_EOL;
    if (count($blueNodes) === 0) {
        return [$cutoff, -1];
    }
    foreach ($blueNodes as $nodeIdx) {
        if ($nofPossibleMerges[$apta[$nodeIdx]['label']] > $cutoff) {
            continue;
        }
        return [$cutoff, $nodeIdx];
    }
    return pickBlueNode($cutoff + 1);
}

function walkit($blueNode, $redNode) {
    global $apta, $sofAlphabet;
    if ($apta[$blueNode]['label'] !== LABEL_NONE) {
        if ($apta[$redNode]['label'] !== LABEL_NONE) {
            if ($apta[$blueNode]['label'] !== $apta[$redNode]['label']) {
                throw new Exception();
            }
        } else {
            setLabel($redNode, $apta[$blueNode]['label']);
        }
    }
    for ($i = 0; $i < $sofAlphabet; ++$i) {
        if ($apta[$blueNode]['children'][$i] !== NO_CHILD) {
            if ($apta[$redNode]['children'][$i] !== NO_CHILD) {
                walkit($apta[$blueNode]['children'][$i], $apta[$redNode]['children'][$i]);
            } else {
                setChild($redNode, $i, $apta[$blueNode]['children'][$i]);
            }
        }
    }
}

function tryMerge($blueNode, $redNode) {
    setFatherPoint($blueNode, $redNode);
    try {
        walkit($blueNode, $redNode);
    } catch (Exception $exception) {
        echo '[MERGE FAILED]' . PHP_EOL;
        return false;
    }
    return true;
}

function exbarSearch() {
    global $apta, $maxNofRedNodes, $changesHistory;
    echo '[EXBAR] No. red nodes: ' . getNofRedNodes() . ', max. no. red nodes: ' . $maxNofRedNodes . PHP_EOL;
    if (getNofRedNodes() > $maxNofRedNodes) {
        echo '[LIMIT OF RED NODES EXCEEDED]' . PHP_EOL;
        return;
    }
    updateNofPossibleMerges();
    list($cutoff, $blueNode) = pickBlueNode();
    echo '[BLUE NODE] The picked node: ' . $blueNode . PHP_EOL;
    if ($blueNode === -1) {
        throw new Exception();
    }
    $redNodes = array_filter($apta, function ($node) { return $node['color'] === COLOR_RED; });
    foreach (array_keys($redNodes) as $redNode) {
        echo '[MERGE] The red node: ' . $redNode . PHP_EOL;
        $nofChangesBefore = count($changesHistory);
        if (tryMerge($blueNode, $redNode)) {
            echo '[MERGE COMPLETED]' . PHP_EOL;
            setColor($blueNode, COLOR_VIOLET);
            exbarSearch();
        }
        $nofChangesAfter = count($changesHistory);
        revertChanges($nofChangesAfter - $nofChangesBefore);
    }
    echo '[RECOLOR] The new red node: ' . $blueNode . PHP_EOL;
    setColor($blueNode, COLOR_RED);
    exbarSearch();
    revertChanges(1);
}

function exbarSearch2() {
    global $apta, $maxNofRedNodes, $changesHistory;
    echo '[EXBAR] No. red nodes: ' . getNofRedNodes() . ', max. no. red nodes: ' . $maxNofRedNodes . PHP_EOL;
    if (getNofRedNodes() > $maxNofRedNodes) {
        echo '[LIMIT OF RED NODES EXCEEDED]' . PHP_EOL;
        return;
    }
    $redNodes = [];
    $blueNodes = [];
    foreach ($apta as $nodeIdx => $node) {
        if ($node['color'] === COLOR_RED) {
            $redNodes[$nodeIdx] = $nodeIdx;
            foreach ($node['children'] as $blueNode) {
                if ($blueNode !== NO_CHILD && $apta[$blueNode]['color'] === COLOR_NONE) {
                    $blueNodes[$blueNode] = $blueNode;
                }
            }
        }
    }
    if (count($blueNodes) === 0) {
        throw new Exception();
    }
    // @todo: sorting blue nodes
    foreach ($blueNodes as $blueNode) {
        echo '[BLUE NODE] The picked node: ' . $blueNode . PHP_EOL;
        foreach ($redNodes as $redNode) {
            echo '[MERGE] The red node: ' . $redNode . PHP_EOL;
            $nofChangesBefore = count($changesHistory);
            if (tryMerge($blueNode, $redNode)) {
                echo '[MERGE COMPLETED]' . PHP_EOL;
                setColor($blueNode, COLOR_VIOLET);
                exbarSearch2();
            }
            $nofChangesAfter = count($changesHistory);
            revertChanges($nofChangesAfter - $nofChangesBefore);
        }
        echo '[RECOLOR] The new red node: ' . $blueNode . PHP_EOL;
        setColor($blueNode, COLOR_RED);
        exbarSearch2();
        revertChanges(1);
    }
}

$start = microtime(true);
while (true) {
    try {
        echo '[RESET] Max. of red nodes: ' . $maxNofRedNodes . PHP_EOL;
        exbarSearch();
        $maxNofRedNodes++;
    } catch (Exception $exception) {
        echo '[SOLUTION HAS BEEN FOUND]' . PHP_EOL;
        echo '- no. nodes before: ' . count($apta) . PHP_EOL;
        echo '- no. nodes after: ' . getNofRedNodes() . PHP_EOL;
        $elapsed = microtime(true) - $start;
        echo '- elapsed time: ' . $elapsed . ' sec.' . PHP_EOL;
        $index = 0;
        $visited = [];
        $map = [];
        $acceptedStates = [];
        $rejectedStates = [];
        indexNodes(0, $index, $visited, $map, $acceptedStates, $rejectedStates);
        $result = [
            'nofNodes' => $index,
            'map' => $map,
            'acceptedStates' => $acceptedStates,
            'rejectedStates' => $rejectedStates
        ];
        echo getDfa($result);
        break;
    }
}

echo '[VERIFICATION]' . PHP_EOL;
$hasIncorrect = false;
foreach (array_slice($lines, 1) as $exampleIdx => $example) {
    $current = 0;
    for ($i = 0; $i < $example[1]; ++$i) {
        $current = $apta[$current]['children'][$example[$i + 2]];
    }
    $isCorrect =
        ($example[0] === 1 && $apta[$current]['label'] === LABEL_ACCEPTED)
        || ($example[0] === 0 && $apta[$current]['label'] === LABEL_REJECTED);
    if ($isCorrect === false) {
        echo '[INCORRECT] The example #' .  $exampleIdx . ' is incorrect' . PHP_EOL;
        $hasIncorrect = true;
    }
}
if ($hasIncorrect === false) {
    echo '- ok' . PHP_EOL;
}
